### PR TITLE
SNOW-1208807 [Local Testing] Support importing non Python file in UDF and access from snowflake_import_directory

### PR DESCRIPTION
**Pull request by sfc-gh-stan**
Please answer these questions before submitting your pull requests. Thanks!

1. What GitHub issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   Fixes SNOW-1208807

2. Fill out the following pre-review checklist:

   - [x] I am adding a new automated test(s) to verify correctness of my new code
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency

3. Please describe how your code solves the related issue.

   This PR supports non Python imports in UDF by creating a temporary directory per UDF execution. All imported files will be copied to this directory, users can access these files by reading the value of `sys._xoptions.get("snowflake_import_directory")` .
   Example usage in official documentation: https://docs.snowflake.com/en/developer-guide/udf/python/udf-python-examples#reading-a-statically-specified-file-using-imports

### CODE REVIEW

File: src/snowflake/snowpark/mock/_plan.py [hunk 1]
```diff
@@ -443,22 +468,32 @@ def handle_udf_expression(
     res.name = quote_name(f"{exp.udf_name}({', '.join(input_data.columns)})".upper())
 
     if udf_name in analyzer.session.udf._udf_level_imports:
+        # Remove UDF level imports
         for module_path in analyzer.session.udf._udf_level_imports[udf_name]:
             if (
                 module_path in sys.path
-            ):  # We need this additional check since children expression might temper with sys.path as well, this needs some future work
+            ):  # TODO: For now we need this additional check since children expression can write to sys.path as well
```
Comment (sfc-gh-jrose) [inline src/snowflake/snowpark/mock/_plan.py:475]: When will this no longer be the case?

File: src/snowflake/snowpark/mock/_plan.py [hunk 2]
```diff
@@ -411,26 +414,48 @@ def handle_udf_expression(
     if udf_name not in udf_registry:
         raise SnowparkSQLException(f"[Local Testing] udf {udf_name} does not exist.")
 
+    # Initialize import directory
```
Comment (sfc-gh-jrose) [inline src/snowflake/snowpark/mock/_plan.py:417]: If a udf is called multiple times in the same query will the modules be copied for each invocation?

File: src/snowflake/snowpark/mock/_plan.py (hunk 1 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/_plan.py:475]: When we work on adding more isolation, e.g. run UDF in a separate process.

File: src/snowflake/snowpark/mock/_plan.py (hunk 2 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/_plan.py:417]: Yes, we could potentially optimize by initializing the import directory at UDF registration.

File: src/snowflake/snowpark/mock/_plan.py [hunk 3]
```diff
@@ -411,26 +414,48 @@ def handle_udf_expression(
...
+    temporary_import_path = tempfile.TemporaryDirectory()
+    analyzer.session.udf._udf_import_directories[udf_name] = temporary_import_path
+    last_import_directory = sys._xoptions.get("snowflake_import_directory")
+    sys._xoptions["snowflake_import_directory"] = temporary_import_path.name
```
Comment (sfc-gh-jrose) [inline src/snowflake/snowpark/mock/_plan.py:421]: I think this should be handled with try/finally or a context manager.

File: src/snowflake/snowpark/mock/_plan.py [hunk 4]
```diff
@@ -443,32 +454,56 @@ def cleanup_imports():
...
             del sys.modules[key]
 
+        # Cleanup import directory
+        analyzer.session.udf._udf_import_directories[udf_name].cleanup()
+
+        # Restore snowflake_import_directory
+        if last_import_directory is not None:
+            sys._xoptions["snowflake_import_directory"] = last_import_directory
+        else:
+            del sys._xoptions["snowflake_import_directory"]
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/_plan.py:469]: what is the `Restore` process here?

File: src/snowflake/snowpark/mock/_plan.py [hunk 5]
```diff
@@ -430,10 +433,18 @@ def handle_udf_expression(
     if udf_name not in udf_registry:
         raise SnowparkSQLException(f"[Local Testing] udf {udf_name} does not exist.")
 
+    # Initialize import directory
+    temporary_import_path = tempfile.TemporaryDirectory()
+    analyzer.session.udf._udf_import_directories[udf_name] = temporary_import_path
+    last_import_directory = sys._xoptions.get("snowflake_import_directory")
```
Comment (sfc-gh-aling) [inline src/snowflake/snowpark/mock/_plan.py:439]: do we actually use `last_import_directory`, is the settable by user?

**Review by sfc-gh-aling — COMMENTED** [review]
only having some questions about the `last_import_directory` part, other parts look good to me

File: src/snowflake/snowpark/mock/_plan.py (hunk 4 above)
Comment (sfc-gh-stan) [reply src/snowflake/snowpark/mock/_plan.py:469]: We need `last_import_directory` when a UDF's input is another UDF's results. On line 510, we calculate the child expression as function input, when this involves another UDF expression, the child UDF should be computed with the corresponding import directory but after its execution this should be restored to the last UDF execution on the stack, so we need to save `last_import_directory` in each `handle_udf_expression` call and restore it at the end.

Comment (sfc-gh-aling) [reply src/snowflake/snowpark/mock/_plan.py:469]: I see, that makes more sense to me now, thanks for the explanation.

**Review by sfc-gh-aling — APPROVED** [review]

**Pull request merged**